src/rest_vol_object.c

```diff
@@ -578,10 +746,368 @@ RV_object_specific(void *obj, const H5VL_loc_params_t *loc_params, H5VL_object_s
+                                "H5Oiterate/H5Ovisit (_by_name) user callback failed for target object ");
+            }
+            else if (callback_ret > 0) {
+                FUNC_GOTO_DONE(callback_ret);
+            }
+
+            /* Make GET request to server */
+
+            /* Setup the host header */
+            host_header_len = strlen(object_iter_data.iter_obj_parent->domain->u.file.filepath_name) +
+                              strlen(host_string) + 1;
+            if (NULL == (host_header = (char *)RV_malloc(host_header_len)))
+                FUNC_GOTO_ERROR(H5E_LINK, H5E_CANTALLOC, FAIL,
+                                "can't allocate space for request Host header");
+
+            strcpy(host_header, host_string);
+
+            curl_headers = curl_slist_append(
+                curl_headers,
+                strncat(host_header, object_iter_data.iter_obj_parent->domain->u.file.filepath_name,
+                        host_header_len - strlen(host_string) - 1));
+
+            /* Disable use of Expect: 100 Continue HTTP response */
+            curl_headers = curl_slist_append(curl_headers, "Expect:");
+
+            if (CURLE_OK != curl_easy_setopt(curl, CURLOPT_HTTPHEADER, curl_headers))
+                FUNC_GOTO_ERROR(H5E_LINK, H5E_CANTSET, FAIL, "can't set cURL HTTP headers: %s", curl_err_buf);
+            if (CURLE_OK != curl_easy_setopt(curl, CURLOPT_HTTPGET, 1))
+                FUNC_GOTO_ERROR(H5E_LINK, H5E_CANTSET, FAIL, "can't set up cURL to make HTTP GET request: %s",
+                                curl_err_buf);
+
+            if (CURLE_OK != curl_easy_setopt(curl, CURLOPT_URL, request_url))
+                FUNC_GOTO_ERROR(H5E_LINK, H5E_CANTSET, FAIL, "can't set cURL request URL: %s", curl_err_buf);
+
+#ifdef RV_CONNECTOR_DEBUG
+            printf("-> Retrieving all links in group using URL: %s\n\n", request_url);
+
+            printf("   /**********************************\\\n");
+            printf("-> | Making GET request to the server |\n");
+            printf("   \\**********************************/\n\n");
+#endif
+
+            CURL_PERFORM(curl, H5E_LINK, H5E_CANTGET, FAIL);
+
+            switch (iter_object_type) {
+                case H5I_FILE:
+                case H5I_GROUP:
+                    if (RV_parse_response(response_buffer.buffer, &object_iter_data, NULL,
+                                          RV_object_iter_callback) < 0)
+                        FUNC_GOTO_ERROR(H5E_LINK, H5E_CANTGET, FAIL, "can't iterate over links");
+                    break;
+
+                case H5I_DATASET:
+                case H5I_DATATYPE:
+                    /* Execute callback on object */
+
+                    if (RV_parse_response(response_buffer.buffer, NULL, &oinfo, RV_get_object_info_callback) <
+                        0)
+                        FUNC_GOTO_ERROR(H5E_LINK, H5E_CANTGET, FAIL, "can't iterate over links");
+
+                    if (!iter_object_name)
```

Comment:
A few issues here:

- It looks like `iter_object_name` never gets set, although I might have missed it
- This looks like it will call the supplied callback twice on dataset and datatype objects if it manages to get past this check
- If the object is a dataset or datatype object, there's no need to make a curl request to try to get all the links from the object